docs/standard/base-types/anchors-in-regular-expressions.md

```diff
@@ -80,18 +80,18 @@ Anchors, or atomic zero-width assertions, specify a position in the string where
 
 ## End of String or Before Ending Newline: \Z  
 
- The `\Z` anchor specifies that a match must occur at the end of the input string, or before `\n` at the end of the input string. It is identical to the `$` anchor, except that `\Z` ignores the <xref:System.Text.RegularExpressions.RegexOptions.Multiline?displayProperty=nameWithType> option. Therefore, in a multiline string, it can only match the end of the last line, or the last line before `\n`.  
+ The `\Z` anchor specifies that a match must occur at the end of the input string, or before `\n` at the end of the input string. It is identical to the `$` anchor, except that `\Z` ignores the <xref:System.Text.RegularExpressions.RegexOptions.Multiline?displayProperty=nameWithType> option. Therefore, in a multiline string, it can only be satisfied by the end of the last line, or the last line before `\n`.  
   
- Note that `\Z` matches `\n` but does not match `\r\n` (the CR/LF character combination). To match CR/LF, include `\r?\Z` in the regular expression pattern.  
+ Note that `\Z` is satisfied by `\n` but is not satisfied by `\r\n` (the CR/LF character combination). To treat CR/LF as if it were `\n`, include `\r?\Z` in the regular expression pattern. Note that this will make the `\r` part of the match - if you are accessing the match, and do not want `\r` to be part of it, you may replace `\r?\Z` with a positive lookahead assert, `(?=\r\n\z|\z|(?<!\r)(?=\n\z))`. In this expression, `\z` asserts the end of the input string.
```

Comment:
The assumption is that we want to recognize `\r\n` but not treat it like two lines. 

For example given "abc\n" in SingleLine mode the pattern ".\Z" will match at position 2 (before the newline) and position 3 (before the end). Now take "abc\r\n". If we want `\r\n` to be treated like a newline, then we want to match at 2 (before newline) and 4 (before end)

`(?=\r\n\z|\z|(?<!\r)(?=\n\z))` does this but `(?=\r\n\z|\n\z|\z)` and `(?=(?:\r?\n)?\z)` will add a third match at position 3.

Reply:
Recommending lookarounds is problematic because lookarounds don't work with non-backtracking, nevermind they're being complicated. It'd be better, for example, to recommend other patterns that would match the newlines and then have a capture group for just the piece you want.

Reply:
That would be a more extended discussion. I think there is value in including this subexpression for reference and for those that need it. Mind if I leave as is for now? I feel like what we have here is a net improvement and I can do more to the docs another day.

Adding support for AnyNewLine is the most useful thing we can do here.

Reply:
If you recommend such patterns then you need to call out that they will prevent NonBacktracking from working / will fail with NonBacktracking.

Reply:
> Adding support for AnyNewLine is the most useful thing we can do here.

How will we achieve that in a way that doesn't use lookarounds?

Reply:
It would not be supported on that engine unless and until it supports lookarounds. I believe @veanes mentioned he had an idea how to do it, so perhaps that will be supported one day.

Reply:
The discussed lookarounds solution for non-backtracking is to run the lookaround pattern at every position. That can lead to quadratic behavior in the length of the input and really complicates the messaging around the option.

Reply:
We can simply disable the option for non backtracking engine if we don't have a way to achieve it with reasonable performance characteristics.

Reply:
I'm not excited at the idea of adding a new feature that doesn't work with non-backtracking. What do folks do here with RE2? With Rust?

Reply:
Neither of those engines support atomic grouping or lookarounds, nor do they support \R. 

Amongst engines that do support backtracking it seems to be fairly common to support `\r\n` as an atomic unit, and \R, more or less how the Unicode standard says. That's true for PCRE, Perl, Java, oniguruma. Whether they do that internally through lowering of some kind, or special casing in the engine, I don't know, but since they support atomic group and lookarounds they can do lowering.

Reply:
> Amongst engines that do support backtracking it seems to be fairly common to support \r\n as an atomic unit

Can you elaborate? What is the syntax someone uses with PCRE for example to achieve the "treat \n or \r\n as a newline" semantic?

Reply:
\R consumes a "generic line break". Exactly what \R that is can  be adjusted. Similarly, what $ and \Z consider a line break can be adjusted, at compile time or within the pattern.

https://www.pcre.org/current/doc/html/pcre2pattern.html#newlines https://www.pcre.org/current/doc/html/pcre2pattern.html#newlineseq
https://www.pcre.org/current/doc/html/pcre2api.html#newlines

eg \R might be equivalent to `(?>\r\n|\n|\x0b|\f|\r|\x85)`. And you might have `(*ANYCRLF)` in your pattern to mean that you want to recognize any of CR, LF, or CRLF.

It makes clear that if CRLF is recognized as a newline, then it will be treated atomically:

> When the newline convention (see ["Newline conventions"](https://www.pcre.org/current/doc/html/pcre2pattern.html#newlines) below) recognizes the two-character sequence CRLF as a newline, this is preferred, even if the single characters CR and LF are also recognized as newlines. For example, if the newline convention is "any", a multiline mode circumflex matches before "xyz" in the string "abc\r\nxyz" rather than after CR, even though CR on its own is a valid newline. (It also matches at the very start of the string, of course.)

Reply:
In Perl, \R matches generic newline: it's not clear whether you can change what $ recognizes.

In Java, $ and \Z and . as appropriate recognize a variety of line terminators including the atomic `\r\n`. But there's an option UNIX_LINES to make them only recognize `\n`. In other words, they have AnyNewLine as proposed, but on by default.
https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html (see Line terminators)

Reply:
@veanes / @olsaarik, I assume for the lookarounds support we discussed, we could by default only permit lookarounds with a max length limited by the pattern rather than by the input, eg no unbounded loops. That would enable the patterns discussed here.

Reply:
Right. So a straightforward fragment is essentially the one where unbounded loops do not occur, perhaps even stricter -- no loops (either bounded or nonbounded). One other issue is that in general NFA mode might still be needed during lookaraounds, but that should be orthogonal to the rest and happen automatically. My understanding is also that all captures are disabled during lookarounds and not stored.